evap/results/tools.py

```diff
@@ -171,42 +199,78 @@ def _calculate_results_impl(course):
+        grouped_results[contributor].extend(results)
+
+    contributor_averages = []
+    contributor_distributions = []
+
+    for contributor in grouped_results:
+        grade_question_results = [result for result in grouped_results[contributor] if result.question.is_grade_question]
+        grade_questions_average = avg([result.average for result in grade_question_results])
+        grade_questions_counts = sum_up_result_counts(grade_question_results)
+        non_grade_question_results = [result for result in grouped_results[contributor] if result.question.is_non_grade_rating_question]
+        non_grade_questions_average = avg([result.average for result in non_grade_question_results])
+        non_grade_questions_counts = sum_up_result_counts(non_grade_question_results)
+        if contributor:
+            # collect the grade question average, non-grade question average and the distributions for each contributor
+            contributor_averages.append(mix(grade_questions_average, non_grade_questions_average, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+            grade_questions_distribution = normalize_distribution(grade_questions_counts)
+            non_grade_questions_distribution = normalize_distribution(non_grade_questions_counts)
+            contributor_distributions.append(mix_distributions(grade_questions_distribution, non_grade_questions_distribution, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+        else:
+            # calculate the grade question average, non-grade question average and the distributions for the course
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = normalize_distribution(grade_questions_counts)
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = normalize_distribution(non_grade_questions_counts)
+
+    # calculate the average grade and distribution over all contributors
+    contributors_average = avg(contributor_averages)
+    contributors_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    for distribution in contributor_distributions:
+        for answer, value in distribution.items():
+            contributors_distribution[answer] += value
+    contributors_distribution = normalize_distribution(contributors_distribution)
+
+    # calculate to which percentage each group influences the total grade
+    contributions_percentage = settings.CONTRIBUTIONS_PERCENTAGE if contributors_average is not None else 0
+    course_grade_questions_percentage = settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_grade_questions_average is not None else 0
+    course_non_grade_questions_percentage = 1 - settings.CONTRIBUTIONS_PERCENTAGE - settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_non_grade_questions_average is not None else 0
+    if contributors_average or course_grade_questions_average or course_non_grade_questions_average:
```

Comment:
how can all of them be zero?

Reply:
if there are no answers yet for a course

Reply:
check